lobbyboy/provider.py

```diff
@@ -1,89 +1,166 @@
             channel: Note that the channel can be None.
                      If called from server_killer, channel will be None.
                      if called when user logout from server, channel is active.
-        """
-        pass
 
-    def ssh_server_command(self, server_id, server_ip):
-        """
-        Args:
-           server_id: the server ssh to, which is returned by you from ``new_server``
-           server_ip: ip or domain name.
         Returns:
-            list: a command in list format, for later to run exec.
+            bool: True if destroy successfully, False if not.
         """
-        pass
-
-    def get_bill(self):
-        pass
-
-    def send_timepass(self, chan, stop_event):
-        start = time.time()
+        ...
 
-        def _print_time_elaspe():
-            while not stop_event.is_set():
-                current = time.time()
-                chan.send(
-                    "\rCreating new server... ({:.1f}s)".format(
-                        current - start
-                    ).encode()
-                )
-            chan.send(b"\r\n")
+    def collection_ssh_keys(self, generate: bool = True, save_path: Path = None) -> List[str]:
+        ssh_keys = self.provider_config.extra_ssh_keys[::] or []
+        if generate:
+            _, pub_key = confirm_ssh_key_pair(save_path=save_path or self.workspace)
+            ssh_keys.append(pub_key)
+        return ssh_keys
 
-        t = threading.Thread(target=_print_time_elaspe)
-        t.start()
+    def default_private_key_path(self, workspace: Path = None, key_type: KeyTypeSupport = KeyTypeSupport.RSA) -> Path:
+        workspace = workspace or self.workspace
+        return workspace.joinpath(f".ssh/id_{key_type.key.lower()}")
 
-    def choose_option(self, ask_prompt, options, channel):
+    def ssh_server_command(self, meta: LBServerMeta, pri_key_path: Path = None) -> List[str]:
         """
-        Utils function.
-        Give the user a list of choices, and return the user choosed one.
-
         Args:
-            ask_prompt: str, a prompt to tell user what they are choosing
-            options: list, a list of string
-            channel: the channel to send and read user input
+           meta: LBServerMeta
+           pri_key_path: path to private key
 
         Returns:
-            string: user choosed option (in options)
-            int: user input number
-
-        Raises:
-            lobbyboy.exceptions.UserCancelException: User press Ctrl-C to cancel the input
+            str: ssh command to connect to provider's server.
         """
-        logger.info(
-            "ask user to choose from {}, propmpt: {}".format(options, ask_prompt)
-        )
-        result = choose_option(ask_prompt, options, channel)
-        logger.info("user choose reuslt: {}".format(result))
-        return result
+        _pri_key_path = pri_key_path or self.default_private_key_path(meta.workspace)
```

Comment:
the `vagrant` ssh command was broken.

`vagrant`'s ssh command is like this:

```
sh -F {data_dir}/vagrant/lobbyboy-1/ssh_config lobbyboy-1
```

This method was left to `abstract` intentionally, the provider's writer should define the ssh command to access servers.

Even though it is the same with `ssh_server_command`, I think it still needs to be explicitly defined.

https://github.com/laixintao/lobbyboy#write-your-own-providers

I prefer to copy this `ssh_server_command` into `digitalocean` and `linode`'s provider and let `vagrant` have a different one, what do you think?

Reply:
of course, user can control this in connection process.
but most of the scenes using lobbyboy are managing remote AWS/Linode/Vulter/... just like 'standard services😂, other like docker/multipass/lima also can be access by common ssh commands, ssh command in `BaseProvider has covered most scenes. why do users need to write them again? i'd like they write less and get more, if users need to control, they only need to overwrite

Reply:
ok, can you fix the vagrant ssh command? then I think I have no further comments. :)

I will document this in readme later on.

Reply:
yes sure👻